server/src/main/java/org/elasticsearch/index/mapper/SourceFieldMapper.java

```diff
@@ -145,7 +143,19 @@ private SourceFieldMapper(boolean enabled, String[] includes, String[] excludes)
         this.includes = includes;
         this.excludes = excludes;
         final boolean filtered = CollectionUtils.isEmpty(includes) == false || CollectionUtils.isEmpty(excludes) == false;
-        this.filter = enabled && filtered ? XContentMapValues.filter(includes, excludes) : null;
+        if (enabled && filtered) {
+            this.parserConfig = XContentParserConfiguration.EMPTY.withFiltering(Set.of(includes), Set.of(excludes));
+            this.filter = (sourceBytes, contentType) -> {
+                BytesStreamOutput streamOutput = new BytesStreamOutput(Math.min(1024, sourceBytes.length()));
+                XContentBuilder builder = new XContentBuilder(XContentType.JSON.xContent(), streamOutput);
+                XContentParser parser = XContentType.JSON.xContent().createParser(parserConfig, sourceBytes.streamInput());
+                builder.copyCurrentStructure(parser);
+                return BytesReference.bytes(builder);
+            };
+        } else {
+            this.parserConfig = null;
+            this.filter = null;
```

Comment:
I think I'd prefer to define `this.filter` as `(sourceBytes, contentType) -> sourceBytes` here, which means we can avoid the null check in `applyFilters`

Reply:
Thanks